test/instruction/instructions/compare/cmp.test.ts

```diff
@@ -0,0 +1,237 @@
+    expect(opcode[0].toBinaryString()).toEqual('0010100000000000')
+  })
+
+  it('should create correct opcode for CMP R7, #255', () => {
+    let registerArray = ['R7', '#255']
+    let opcode = cmpImmInstruction.encodeInstruction(registerArray)
+    expect(opcode[0].toBinaryString()).toEqual('0010111111111111')
+  })
+
+  it('should throw error for CMP R8, #0 because of high register', () => {
+    let registerArray = ['R8', '#0']
+    expect(() => cmpImmInstruction.encodeInstruction(registerArray)).toThrow()
+  })
+
+  it('should throw error for CMP R1, R2, R3 because of too many params', () => {
+    let registerArray = ['R1', 'R2', '#0']
+    expect(() => cmpImmInstruction.encodeInstruction(registerArray)).toThrow()
+  })
+})
+
+describe('test executeInstruction function for CMP with immediate', () => {
+  it('should return correct result for CMP R1, #2; R1 = 0x01', () => {
+    let registerArray = ['R1', '#2']
+    let opcode = cmpImmInstruction.encodeInstruction(registerArray)
+    cmpImmInstruction.executeInstruction(opcode, registers, memory)
+    expect(registers.readRegister(Register.R1).value).toEqual(0x01)
+    expect(registers.isFlagSet(Flag.N)).toBeTruthy()
+    expect(registers.isFlagSet(Flag.Z)).toBeFalsy()
+    expect(registers.isFlagSet(Flag.C)).toBeFalsy()
+    expect(registers.isFlagSet(Flag.V)).toBeFalsy()
+  })
+
+  it('should return correct result for CMP R2, #1; R2 = 0x02', () => {
+    let registerArray = ['R2', '#1']
+    let opcode = cmpImmInstruction.encodeInstruction(registerArray)
+    cmpImmInstruction.executeInstruction(opcode, registers, memory)
+    expect(registers.readRegister(Register.R2).value).toEqual(0x02)
+    expect(registers.isFlagSet(Flag.N)).toBeFalsy()
+    expect(registers.isFlagSet(Flag.Z)).toBeFalsy()
+    expect(registers.isFlagSet(Flag.C)).toBeTruthy()
+    expect(registers.isFlagSet(Flag.V)).toBeFalsy()
+  })
+
+  it('should return correct result for CMP R1, #255; R1 = 0x01', () => {
+    let registerArray = ['R1', '#255']
+    let opcode = cmpImmInstruction.encodeInstruction(registerArray)
+    cmpImmInstruction.executeInstruction(opcode, registers, memory)
+    expect(registers.readRegister(Register.R1).value).toEqual(0x01)
+    expect(registers.isFlagSet(Flag.N)).toBeTruthy()
+    expect(registers.isFlagSet(Flag.Z)).toBeFalsy()
+    expect(registers.isFlagSet(Flag.C)).toBeFalsy()
+    expect(registers.isFlagSet(Flag.V)).toBeFalsy()
+  })
+
+  it('should return correct result for CMP R3, #1; R3 = 0x80000000', () => {
+    let registerArray = ['R3', '#1']
+    let opcode = cmpImmInstruction.encodeInstruction(registerArray)
+    cmpImmInstruction.executeInstruction(opcode, registers, memory)
+    expect(registers.readRegister(Register.R3).value).toEqual(0x80000000)
+    expect(registers.isFlagSet(Flag.N)).toBeFalsy()
+    expect(registers.isFlagSet(Flag.Z)).toBeFalsy()
+    expect(registers.isFlagSet(Flag.C)).toBeTruthy()
+    expect(registers.isFlagSet(Flag.V)).toBeTruthy()
+  })
```

Comment:
I would recommend to also make tests with high values like 0x7fffffff similar like for cmn. Even if it is not a normal situation, we should check with the tests if the results will be as on the physical board.

Reply:
@taggatips this one is for you.

Reply:
What kind of comparison would you like to happen (second param)? From my point of view, the given tests account for most of the possible main results of subtraction ...